src/lib/logger.js

```diff
@@ -0,0 +1,143 @@
+  warn: uninitializedLogger,
+  debug: uninitializedLogger,
+};
+
+/**
+ * Note that all the methods use arrow functions to bind 'this', this is so we
+ * can pass them to the logger object to be called outside of the immediate
+ * context of the Logger instance
+ */
+class Logger {
+  constructor(isClient, displayAlert) {
+    if (isClient === undefined) {
+      throw new Error(
+        'The logger constructor needs an isClient parameter supplied',
+      );
+    }
+    if (isClient && !displayAlert) {
+      throw new Error(
+        'on the client the logger constructor needs a displayAlert function supplied',
+      );
+    }
+    this.isClient = isClient;
+    this.displayAlert = displayAlert;
+  }
+
+  /**
+   * Logs a fatal error. This should only be called if an error occurred that
+   * completely breaks our system and requires a complete system reboot.
+   * On the client it will refresh the page after displaying an alert,
+   * on the server it will exit the server which will make forever restart
+   * the server.
+   * @param {Error | string} err - The fatal error to be logged, either a string or
+   * an Error object
+   */
+  fatal = async err => {
+    const errorObject = err instanceof Error ? err : new Error(err);
+    if (this.isClient) {
+      // We are actually lying currently, as developers haven't been notified
+      // when we hook up sentry we will be notifying developers though
+      await this.displayAlert(
+        'A fatal error has occurred, and the developers' +
+          ' have been notified. We apologize for the inconvenience, and we are afraid' +
+          ' that we will have to reload the page now',
+      );
+      window.location.reload(true);
+    } else {
+      // eslint-disable-next-line no-console
+      console.error(
+        'A fatal error occured, and the server is exiting. See the error below:',
+      );
+      // eslint-disable-next-line no-console
+      console.error(errorObject);
+      process.exit(1);
+    }
+  };
+
+  /**
+   * Logs an error. This should only be called on unexpected behaviour that should
+   * never happen as long as everything works
+   * @param {Error | string} err - The error to be logged, either a string or
+   * an Error object
+   */
+  error = err => {
+    const errorObject = err instanceof Error ? err : new Error(err);
+    // eslint-disable-next-line no-console
+    console.error(errorObject);
+  };
+
+  /**
+   * Logs a warning. This should be called when an event occurs that is usually unwanted
+   * but under certain circumstances could be okay
+   * @param {Error | string} warning - The warning to be logged, either a string or
+   * an Error object
+   */
+  warn = warning => {
+    const errorObject = warning instanceof Error ? warning : new Error(warning);
+    // eslint-disable-next-line no-console
+    console.warn(errorObject);
+  };
+
+  /**
+   * Logs a debug statement. This can be called whereever the developer thinks
```

Comment:
`wherever`

Reply:
You learn something new every day! Thanks :)

Reply:
:D